### PR TITLE
Make Metaserver changable in settings

### PR DESCRIPTION
**Pull request by d-513**

### CODE REVIEW

**Comment by d-513** [comment]
QUrl#isValid is a bad function and doesn't validate properly, so that's a TODO. I can pass it anything and it will accept it

**Review by txtsd — APPROVED** [review]
LGTM!

File: launcher/ui/pages/global/APIPage.cpp [hunk 1]
```diff
@@ -79,6 +82,20 @@ void APIPage::applySettings()
     s->set("PastebinURL", pastebinURL);
     QString msaClientID = ui->msaClientID->text();
     s->set("MSAClientIDOverride", msaClientID);
+    QUrl metaURL = ui->metaURL->text();
+    if(metaURL.isEmpty())
+    {
+        s->set("MetaURL", BuildConfig.META_URL);
+        return;
+    }
```
Comment (Scrumplex) [inline launcher/ui/pages/global/APIPage.cpp]: Don't like this. Maybe we want to do this like with MSA client ID instead? We register a setting called `MetaURLOverride` and only use it if it is present. If it is empty, we just use the value from BuildConfig.

**Review by Scrumplex — CHANGES_REQUESTED** [review]
See #161

File: launcher/ui/pages/global/APIPage.cpp (hunk 1 above)
Comment (d-513) [reply launcher/ui/pages/global/APIPage.cpp]: Ok

**Review by Scrumplex — APPROVED** [review]

**Comment by Scrumplex** [comment]
Let's merge this once 1.2.1 is out

File: launcher/ui/pages/global/APIPage.ui [hunk 2]
```diff
@@ -147,6 +147,58 @@
...
+           <widget class="QLineEdit" name="metaURL">
+            <property name="placeholderText">
+             <string>(Default)</string>
+            </property>
+           </widget>
+          </item>
+          <item>
+           <widget class="QLabel" name="label_5">
+            <property name="text">
+             <string>Enter a custom URL for meta here. </string>
```
Comment (Scrumplex) [inline launcher/ui/pages/global/APIPage.ui]:
```suggestion
             <string>Enter a custom URL for meta here.</string>
```

File: launcher/ui/pages/global/APIPage.ui [hunk 3]
```diff
@@ -147,6 +147,58 @@
...
+           <widget class="Line" name="line_3">
+            <property name="orientation">
+             <enum>Qt::Horizontal</enum>
+            </property>
+           </widget>
+          </item>
+          <item>
+           <widget class="QLabel" name="label_4">
+            <property name="text">
+             <string>You can set this to a third-party metadata server to use patched libraries or other hacks</string>
```
Comment (Scrumplex) [inline launcher/ui/pages/global/APIPage.ui]:
```suggestion
             <string>You can set this to a third-party metadata server to use patched libraries or other hacks.</string>
```

File: launcher/ui/pages/global/APIPage.ui [hunk 4]
```diff
@@ -147,6 +147,58 @@
...
+         <layout class="QVBoxLayout" name="verticalLayout_5">
+          <item>
+           <widget class="Line" name="line_3">
+            <property name="orientation">
+             <enum>Qt::Horizontal</enum>
+            </property>
+           </widget>
+          </item>
+          <item>
+           <widget class="QLabel" name="label_4">
```
Comment (Scrumplex) [inline launcher/ui/pages/global/APIPage.ui]: This generates a warning on my end. The id `label_4` is already used.
```suggestion
           <widget class="QLabel" name="label_5">
```

File: launcher/ui/pages/global/APIPage.ui [hunk 5]
```diff
@@ -147,6 +147,58 @@
...
+          </item>
+          <item>
+           <widget class="QLineEdit" name="metaURL">
+            <property name="placeholderText">
+             <string>(Default)</string>
+            </property>
+           </widget>
+          </item>
+          <item>
+           <widget class="QLabel" name="label_5">
```
Comment (Scrumplex) [inline launcher/ui/pages/global/APIPage.ui]:
```suggestion
           <widget class="QLabel" name="label_6">
```

**Review by Scrumplex — CHANGES_REQUESTED** [review]
nitpicks

**Comment by Scrumplex** [comment]
btw. Use Qt Designer, it automatically fixes some problems like this and stuff like formatting

**Comment by DioEgizio** [comment]
this is going to conflict with #497 btw

File: launcher/ui/pages/global/APIPage.ui [hunk 6]
```diff
@@ -147,6 +147,58 @@
          </layout>
         </widget>
        </item>
+       <item>
+        <widget class="QGroupBox" name="groupBox_meta">
+         <property name="title">
+          <string>Metadata Server</string>
```
Comment (Scrumplex) [inline launcher/ui/pages/global/APIPage.ui]: Needs mnemonic
```suggestion
          <string>Meta&amp;data Server</string>
```

**Review by Scrumplex — DISMISSED** [review]

**Comment by Scrumplex** [comment]
rebase pls

File: launcher/ui/pages/global/APIPage.cpp [hunk 7]
```diff
@@ -46,6 +46,7 @@
 #include "settings/SettingsObject.h"
 #include "tools/BaseProfiler.h"
 #include "Application.h"
+#include "BuildConfig.h"
```
Comment (flowln) [inline launcher/ui/pages/global/APIPage.cpp]:
```suggestion
```

this doesn't seem to be used anywhere in this file?

**Review by flowln — APPROVED** [review]

**Comment by DioEgizio** [comment]
Rebase on develop wen

**Comment by d-513** [comment]
> Rebase on develop wen

soon, merging ui files is hell please help me

**Review by timoreo22 — APPROVED** [review]

**Comment by DioEgizio** [comment]
> > Rebase on develop wen
> 
> soon, merging ui files is hell please help me

wen

**Comment by d-513** [comment]
> > > Rebase on develop wen
> > 
> > 
> > soon, merging ui files is hell please help me
> 
> wen

build issues

**Comment by DioEgizio** [comment]
arch moment

**Comment by Scrumplex** [comment]
re-review please. I have rebased this on develop

**Comment by DioEgizio** [comment]
> re-review please. I have rebased this on develop

Well, there are 3 approvals, so someone could do the funny anyways

**Comment by Scrumplex** [comment]
> > re-review please. I have rebased this on develop
> 
> Well, there are 3 approvals, so someone could do the funny anyways

Nah. I might have borked something while resolving conflicts

**Comment by Scrumplex** [comment]
bruh, conflicts again

**Comment by DioEgizio** [comment]
conflict hell moments

**Review by Scrumplex — APPROVED** [review]

**Pull request merged**